### PR TITLE
make ResourceLimiter operate on Store data; add hooks for entering and exiting native code

### PR DESCRIPTION
**Pull request by pchickey**
Based on #2897 - draft until that PR lands.

This PR does two things:

* The `ResourceLimiter` trait is now implemented on the data inside a `Store` rather than as a standalone owned type. This change required changing the structure of the Store internals a bunch. Additionally, the trait is now defined in `wasmtime_runtime` and re-exported by `wasmtime`.
* Two new functions, `entering_native_code_hook` and `exiting_native_code_hook`, are added to `Store`. These functions get to operate on the data inside a `Store` much like resource limiter's mut methods. These functions give embedding the opportunity to measure time (or other resources) and trap before entering any native code (i.e. a `Func`) linked into the `Store`, and again before returning from that native code.

### CODE REVIEW

**Review by alexcrichton — COMMENTED** [review]
Ok reading again now that the diff is a bit smaller, this is still looking good. (modulo the `Innermost` which I don't know how to make better and isn't really all that bad).

Can you add some tests thouggh for the enter/exit hooks? Ideally we'd have tests ensuring that they fire for all kinds of created host functions (e.g. `Linker`-defined, `Func`-defined, etc).

Finally, do you think it would make sense to run the hooks around the initial entry point into wasm as well? (e.g. the single entry point into wasm we have in `func.rs`).

File: tests/all/module_linking.rs [hunk 1]
```diff
@@ -253,8 +253,8 @@ fn limit_memories() -> Result<()> {
             )
         "#,
     )?;
-    let mut store = Store::new(&engine, ());
-    store.limiter(StoreLimitsBuilder::new().memories(10).build());
+    let mut store = Store::new(&engine, StoreLimitsBuilder::new().memories(10).build());
+    store.limiter(|s| s as &mut dyn ResourceLimiter);
```
Comment (alexcrichton) [inline tests/all/module_linking.rs:257]: Is the `as &mut dyn ResourceLimiter` required here in that rustc can't infer it?

File: crates/wast/src/wast.rs [hunk 2]
```diff
@@ -54,9 +54,9 @@ impl<T> Outcome<T> {
     }
 }
 
-impl WastContext {
+impl<T> WastContext<T> {
```
Comment (alexcrichton) [inline crates/wast/src/wast.rs:57]: Oh how come this change was necessary? I didn't think we even gave access to the `Store`...

File: tests/all/module_linking.rs (hunk 1 above)
Comment (bjorn3) [reply tests/all/module_linking.rs:257]: `|s| s` doesn't attempt to coerce `s` AFAIK.

File: crates/wast/src/wast.rs (hunk 2 above)
Comment (pchickey) [reply crates/wast/src/wast.rs:57]: This is used over in the fuzzing crate, which uses a resource limiter. So, while previously the store could have a `()` data and also own a `dyn ResourceLimiter`, now the store has to contain the `dyn ResourceLimiter`.

**Comment by alexcrichton** [comment]
Locally I've got two benchmarks for changes like this:

<details>

<summary>host -> wasm overhead</summary>

```rust
fn host_to_wasm_overhead() -> anyhow::Result<()> {
    use wasmtime::*;
    let engine = Engine::default();
    let module = Module::new(
        &engine,
        r#"
            (module
                (func (export "run") (param i32 i32) (result i32)
                    local.get 0
                    local.get 1
                    i32.add
                )
            )
        "#,
    )?;
    let mut store = Store::new(&engine, ());
    let instance = Instance::new(&mut store, &module, &[])?;
    let run = instance.get_typed_func::<(i32, i32), i32, _>(&mut store, "run")?;
    let now = std::time::Instant::now();
    let total = 100_000_000;
    for _ in 0..total {
        run.call(&mut store, (1, 1))?;
    }
    let time = now.elapsed();
    println!(
        "use Time: {:?}, each:{} ns/op",
        time,
        time.as_nanos() / (total as u128)
    );
    Ok(())
}
```

</details>

<details>

<summary>wasm -> host overhead</summary>

```rust

fn wasm_to_host() -> anyhow::Result<()> {
    use wasmtime::*;
    let engine = Engine::default();
    let module = Module::new(
        &engine,
        r#"
            (module
                (import "" "add" (func $add (param i32 i32) (result i32)))

                (func (export "run") (param i32)
                    loop
                        i32.const 1
                        i32.const 2
                        call 0
                        drop

                        local.get 0
                        i32.const -1
                        i32.add
                        local.tee 0
                        br_if 0
                    end
                ))
        "#,
    )?;
    let mut store = Store::new(&engine, ());
    let func = Func::wrap(&mut store, |a: u32, b: u32| a + b);
    let instance = Instance::new(&mut store, &module, &[func.into()])?;
    let run = instance.get_typed_func::<i32, (), _>(&mut store, "run")?;
    let now = std::time::Instant::now();
    let total = 1_000_000_000;
    run.call(&mut store, total)?;
    let time = now.elapsed();
    println!(
        "use Time: {:?}, each:{} ns/op",
        time,
        time.as_nanos() / (total as u128)
    );
    Ok(())
}
```

</details>

The numbers I got for this PR on my local macOS machine are:

| | before | after |
|---|---|---|
| host -> wasm | 29ns | 30ns |
| wasm -> host | 1ns | 2ns |

So it looks like adding these unconditional branches has pretty negligible overhead. If it *really* matters in some use case we could always add a compile-time option to disable them, but otherwise I don't think there's too much impact.

One thing that's required, though, is to add `#[inline]` to a few more methods. I noticed this when profiling and gathered the numbers with some local changes to this PR. The reason is that `StoreInner` methods which previously had a generic in scope are now concrete on `StoreInnermost` so they require `#[inline]` to get cross-crate inlined. The ones I added inline too are:

* `StoreInnermost::signal_handler`
* `StoreInnermost::vminterrupts`
* `StoreInnermost::async_cx`
* `StoreInnermost::default_callee`
* `StoreInnermost::async_support`
* `StoreInnermost::engine`
* `StoreInnermost::data`
* `StoreInnermost::data_mut`

**Comment by alexcrichton** [comment]
👍

**Pull request merged**